src/enumerable/enumerable_gather.jl

```diff
@@ -0,0 +1,75 @@
+struct EnumerableGather{T,S,F,I,A} <: Enumerable
+    source::S
+    fields::F
+    indexFields::I
+    savedFields::A
+    key::Symbol
+    value::Symbol
+end
+
+struct Not{T}
+    val::T
+end
+
+function gather(source::Enumerable, args...; key::Symbol = :key, value::Symbol = :value)
+    fields = fieldnames(eltype(source))
+    
+    if length(args) > 0
+        indexFields = ()
+        firstArg = true
+        for arg in args
+            if typeof(arg) == Symbol
+                indexFields = (indexFields..., arg)
+            else typeof(arg) == Not{Symbol}
+                if firstArg
+                    indexFields = (a for a in fields if a != arg.val)
+                else
+                    indexFields = (a for a in indexFields if a != arg.val)
+                end
+            end
+            firstArg = false
+        end
+    else
+        indexFields = fields
+    end
+
+    savedFields = (n for n in fields if !(n in indexFields)) # fields that are not in `indexFields`
+    T = NamedTuple{(key, value, savedFields...)}
```

Comment:
We should put the saved fields to the front, then the key and then the value column.

Reply:
And we should construct a concrete `NamedTuple` that includes type information for the fields.